### PR TITLE
Fix handling of function declarations in loops with let/consts

### LINKED ISSUE #1124: Internal error with let inside while loop
This simple code:
```javascript
while (true) {
  let x = null;
  function f() {
    x();
  }
}
```

Gives the following trace (cmd line args: `--language_in=ECMASCRIPT6_STRICT -O SIMPLE`):
```
INTERNAL COMPILER ERROR.
Please report this problem.

Invalid attempt to remove node: NAME $jscomp$loop$0 3 [source_file: bug_closure.js] of CALL 3 [free_call: 1] [source_file: bug_closure.js]
  Node(NAME $jscomp$loop$0): bug_closure.js:3:2
  function f() {
  Parent(CALL): bug_closure.js:3:2
  function f() {

	at com.google.javascript.jscomp.NodeUtil.removeChild(NodeUtil.java:2293)
	at com.google.javascript.jscomp.UnreachableCodeElimination$EliminationPass.removeNode(UnreachableCodeElimination.java:261)
	at com.google.javascript.jscomp.UnreachableCodeElimination$EliminationPass.removeDeadExprStatementSafely(UnreachableCodeElimination.java:251)
	at com.google.javascript.jscomp.UnreachableCodeElimination$EliminationPass.visit(UnreachableCodeElimination.java:104)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:609)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:603)
	at com.google.javascript.jscomp.NodeTraversal.traverseBlockScope(NodeTraversal.java:652)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:597)
	at com.google.javascript.jscomp.NodeTraversal.traverseBlockScope(NodeTraversal.java:652)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:597)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:603)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:603)
	at com.google.javascript.jscomp.NodeTraversal.traverse(NodeTraversal.java:295)
	at com.google.javascript.jscomp.NodeTraversal.traverseEs6(NodeTraversal.java:560)
	at com.google.javascript.jscomp.UnreachableCodeElimination$1.enterFunction(UnreachableCodeElimination.java:81)
```

Note that it works fine with `var` instead of `let`.

### PR DESCRIPTION
**Pull request by Dominator008**
Convert function declaration to variable declaration so that the IIFE can still be applied.

Unmark calls to captured functions like $jscomp$loop.x() as FREE_CALL.

Fixes #1124

### CODE REVIEW

File: test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java [hunk 1]
```diff
@@ -708,6 +708,46 @@ public void testDoWhileForOfCapturedLet() {
             "} while (false);"));
   }
 
+  public void testIssue1124() {
```
Comment (MatrixFrog) [inline test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java]: I think there's a convention that "issue 1124" means 1124 on code.google.com. Let's add a comment with a link so it's unambiguous.

File: src/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclaration.java [hunk 2]
```diff
@@ -330,7 +331,7 @@ private void transformLoopClosure() {
                 // Change declaration to assignment, or just drop it if there's
                 // no initial value.
                 if (reference.hasChildren()) {
-                  declaration = reference.getParent(); // Might have changed now
+                  declaration = referenceParent; // Might have changed now
```
Comment (MatrixFrog) [inline src/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclaration.java]: I don't understand this comment but it seems false.

Comment (Dominator008) [inline src/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclaration.java]: Broke tests, reverted to use reference.getParent() :(

File: test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java (hunk 1 above)
Comment (Dominator008) [inline test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java]: Done

File: src/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclaration.java [hunk 3]
```diff
@@ -343,6 +343,10 @@ private void transformLoopClosure() {
               }
 
               // Change reference to GETPROP.
```
Comment (MatrixFrog) [inline src/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclaration.java]: Now this comment is misplaced.

**Comment by MatrixFrog** [comment]
Can you make sure the pull request description is thorough since that will end up being the CL description internally?

**Comment by Dominator008** [comment]
I assume "pull request description" is just the first comment?

**Comment by MatrixFrog** [comment]
Not sure but I'm seeing this now, which is much better :)

> Convert function declaration to variable declaration so that the IIFE can still be applied.
> 
> Unmark calls to captured functions like $jscomp$loop.x() as FREE_CALL.
> 
> Fixes #1124

**Comment by Dominator008** [comment]
Sounds good :)

**Comment by Dominator008** [comment]
Should we merge this?

File: test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java [hunk 4]
```diff
@@ -708,6 +708,66 @@ public void testDoWhileForOfCapturedLet() {
...
+        LINE_JOINER.join(
+        "var $jscomp$loop$0 = {x:undefined};",
+        "while (true) {",
+        "  $jscomp$loop$0.x = null;",
+        "  var f = function($jscomp$loop$0) {",
+        "    return function() {",
+        "      ($jscomp$loop$0.x)();",
+        "    };",
+        "  }($jscomp$loop$0);",
+        "  $jscomp$loop$0 = {x:$jscomp$loop$0.x};",
```
Comment (MatrixFrog) [inline test/com/google/javascript/jscomp/Es6RewriteBlockScopedDeclarationTest.java]: This seems quite strange but I guess that's not really related to this CL.

**Pull request merged**